src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -28928,6 +28928,244 @@ public void Init(T t)
                 Diagnostic(ErrorCode.WRN_MemberNotNull, "}").WithArguments("field").WithLocation(18, 5));
         }
 
+        [Fact, WorkItem(48996, "https://github.com/dotnet/roslyn/issues/48996")]
+        public void DisallowNullProperty_Constructor()
+        {
+            var source = @"
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+public class C1
+{
+    [DisallowNull, NotNull] public string? F { get; set; }
+
+    public C1() { } // 1
+    public C1(string? F) // 2
+    {
+        this.F = F; // 3
+    }
+
+    public C1(int i)
+    {
+        this.F = i.ToString();
```

Comment:
Why does this not warn given `ToString` returns `string?`

Reply:
perhaps it is overridden on `int`?

Reply:
I'm not sure if it does or does not and that's another reason I think we should avoid `ToString` in our tests here. These tests should be concretely about the behavior of nullable flow. By using `ToString` from the framework we are introducing a bit of an ambiguity as to what we are actually testing. 

It also means when we upgrade to new reference assemblies it could potentially change the meaning of these tests. For most APIs that isn't really an issue because breaks, or changes at all, are rare. With nullability though they are actively annotating so expectation is there should be some change.

Reply:
how about `GetType()`?

Reply:
Same. In general I'd avoid using any BCL method when we're trying to validate nullability. I think it makes the test harder to understand and less resilient to the type of change that we tend to make with reference assembly updates.

Reply:
I feel like this is an argument to avoid using built-in members entirely in nullable tests and that is a fairly onerous ask.

Basically, yes, the risk exists that the annotation will change on the ref assemblies used by tests, and if it does change, then hundreds of tests in here are going to break. I'm a little resistant to the idea that we need to account for it here because it is going to need to be accounted for en masse.

A more reasonable ask would be perhaps to verify what are the annotations on ref assemblies that *are* used in our nullable tests in practice, and ensure that the annotations on ref assemblies for tests do not change accidentally, as well as gain a sense of how we *would* change them, if we wanted to adopt the .NET 5 nullability annotations for certain tests, for example.

Reply:
> I feel like this is an argument to avoid using built-in members entirely in nullable tests and that is a fairly onerous ask.

The majority of tests in this class use the `CreateNullableCompilation` helper to run analysis. That uses either the `netstandard2.0` or the `net451` reference assemblies hence all of the built-in members are oblivious. If the intent is to test oblivious then that works. My suspicion is though that in most cases the intent is not to test oblivious. Perhaps I'm wrong here though. 

> A more reasonable ask would be perhaps to verify what are the annotations on ref assemblies that are used in our nullable tests in practice

The answer to this depends on the test. If the test uses the `netstandard2.0` reference assemblies then there is one answer, if it uses `netcoreapp3.1` then there is a different answer. Basically it is subject to some degree of change. Generally that the API will get stronger though which is a "safe" change from this perspective. 

Consider this specific case though: the test here is using the `netstandard2.0` reference assemblies which means this is validating that `string~` can assign here. Was that the intent? Reading the test I thought the intent was to check what happened when either `string?` or `string!` was the source of the assignment. Am I just reading the intent of this test wrong (or reading the setup code wrong)?

> I'm a little resistant to the idea that we need to account for it here because it is going to need to be accounted for en masse.

I'm not suggesting we go back and rewrite all the tests. At the same time seems like something we could avoid going forward with new ones. 

> A more reasonable ask would be perhaps to verify what are the annotations on ref assemblies that are used in our nullable tests in practice, and ensure that the annotations on ref assemblies for tests do not change accidentally

The model we try to operate on is that we keep pushing the `.NET Core App` reference assemblies forward. Observe that while we have `netcoreapp3.1` assemblies we don't have `netcoreapp3.0` or `netcoreapp2.1`. That's because we're always effectively on latest. The API sets which are stable are `netstandard2.0`, `net451`, etc ...

Reply:
Looking a little more closely, the usages of ToString() in practice are usually either to:

1. demonstrate that "possible null reference receiver" warnings are given where appropriate
2. get a non-null string in order to update a variable's flow state to not-null.

I think there is no risk from continuing to do (1)--annotations on the member won't affect whether such warnings are given.

In these tests it's perhaps more straightforward to achieve (2) by assigning literal strings. So I think I will resolve this thread by changing the new tests to assign literal strings instead.

In practice, there is not going to be a difference between assigning an "oblivious" return value versus a "not annotated" return value--since both will just turn into the "not null" flow state. However, you've raised a good point that the annotations on ToString() in .NET Core make it harder to understand tests here which assign the return of ToString() to something, because the reader is not certain whether such annotations are present in the test or not. So we can try to refrain from doing that in the future.